### ISSUE TITLE
Hydration->cache_store_notices breaks NoticeHandler->convert_notices_to_wp_errors

### ISSUE THREAD

**Issue by lovec741** [woothemes/woocommerce#45446]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When I set add a notice with `wc_add_notice()` and redirect to `wc_get_checkout_url()` the notice doesn't get displayed on the checkout page, but gets displayed when the user tries to pay again.

I believe, this is caused by this series of events:
1. I add a notice to `'wc_notices'` in the WC session and redirect.
2.  `Automattic\WooCommerce\Blocks\Domain\Services\Hydration->cache_store_notices()` gets called and sets `'wc_notices'` in the WC session to `null`
3. `Automattic\WooCommerce\StoreApi\Utilities\NoticeHandler->convert_notices_to_wp_errors()` gets called, but the `'wc_notices'` in the WC session is `null`, so the notice isn't displayed.
4. `Automattic\WooCommerce\Blocks\Domain\Services\Hydration->restore_cached_store_notices()` gets called and restores `'wc_notices'` in the WC session, but too late.

### Expected behavior

The notice gets rendered on the checkout page.

### Actual behavior

The notice doesn't get rendered on the checkout page. It gets rendered after the checkout is submitted again.

### Steps to reproduce

1. Implement a `WC_Payment_Gateway` with a webhook (for example `'woocommerce_api_'.$this->id.'_reply'`)
    - inside the callback add a notice with `wc_add_notice()` and redirect to `wc_get_checkout_url()` 
2. Checkout with the payment gateway.
3. Get redirected to the url matching the webhook from the gateway. 
4. Get redirected by the callback to the checkout page. -> The notice isn't displayed.
5. Try to pay again. -> The notice gets displayed.

<details>
<summary>WordPress Environment</summary>

```
### WordPress Environment ###

WordPress address (URL): http://localhost/php8.1/wordpress6.4.3
Site address (URL): http://localhost/php8.1/wordpress6.4.3
WC Version: 8.6.1
REST API Version: ✔ 8.6.1
WC Blocks Version: ✔ 11.8.0-dev
Action Scheduler Version: ✔ 3.7.1
Log Directory Writable: ✔
WP Version: 6.4.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: cs_CZ
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.52 (Ubuntu)
PHP Version: 8.1.27
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 10000
cURL Version: 7.81.0
OpenSSL/3.0.2

SUHOSIN Installed: –
MySQL Version: 8.0.36-0ubuntu0.22.04.1
Max Upload Size: 2 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.6.1
WC Database Prefix: wp_
Celková velikost databáze: 5.72MB
Velikost dat databáze: 3.69MB
Velikost indexu databáze: 2.03MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
unimodul_transactions: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.06MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.06MB + Index: 0.09MB + Engine InnoDB
wp_gla_attribute_mapping_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_gla_budget_recommendations: Data: 0.19MB + Index: 0.14MB + Engine InnoDB
wp_gla_merchant_issues: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_gla_shipping_rates: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gla_shipping_times: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.08MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.09MB + Index: 0.17MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 11
nav_menu_item: 8
page: 7
post: 2
product: 9
shop_order_placehold: 8
wp_navigation: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ❌
					Váš obchod nepoužívá HTTPS. Zjistěte více o HTTPS a SSL certifikátech.
Hide errors from visitors: ❌Chybová hlášení by neměla být zobrazována návštěvníkům.

### Active Plugins (2) ###

WooCommerce TestModul: Autor: Test – 105.109
WooCommerce: Autor: Automattic – 8.6.1

### Inactive Plugins (3) ###

Akismet Anti-spam: Spam Protection: Autor: Automattic - Anti-spam Team – 5.3.1
Google Listings and Ads: Autor: WooCommerce – 2.6.0
Hello Dolly: Autor: Matt Mullenweg – 1.7.2

### Settings ###

API Enabled: –
Force SSL: –
Currency: CZK (Kč)
Currency Position: right_space
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Obchod: #7 - /obchod/
Košík: #8 - /kosik/
Pokladna: #9 - /pokladna/
Můj účet: #10 - /muj-ucet/
Obchodní podmínky: ❌ Stránka není zvolena

### Theme ###

Name: Twenty Twenty-Three
Version: 1.3
Author URL: https://wordpress.org
Child Theme: ❌ – Pokud upravujete WooCommerce šablonu
kterou jste nevytvořili
doporučujeme použít odvozenou šablonu. Viz Jak vytvořit odvozenou šablonu (anglicky)

WooCommerce Support: ✔

### Templates ###

Overrides: /var/www/html/php8.1/wordpress6.4.3/wp-content/plugins/woocommerce/templates/block-notices/error.php
/var/www/html/php8.1/wordpress6.4.3/wp-content/plugins/woocommerce/templates/block-notices/notice.php
/var/www/html/php8.1/wordpress6.4.3/wp-content/plugins/woocommerce/templates/block-notices/success.php


### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-pre-publish-modal
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2024-03-08 19:54:21 +01:00
Options: ✔
Notes: 70
Onboarding: completed

### Action Scheduler ###

Dokončeno: 117
Oldest: 2024-03-03 20:55:34 +0100
Newest: 2024-03-08 16:56:01 +0100

Čekající: 2
Oldest: 2024-03-08 16:58:04 +0100
Newest: 2024-03-09 15:50:59 +0100


### Status report information ###

Generated at: 2024-03-08 16:57:14 +01:00
```

<details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by jorgeatorres** [comment]
Hi @lovec741,

Thanks for reaching out. Can you confirm if this is also happening when the legacy (shortcode-based) checkout is used? Or is this a Blocks-only issue? Thanks!

**Comment by lovec741** [comment]
Hello @jorgeatorres,

I can confirm this is a Blocks-only issue. 
When I switch to shortcodes [per the documentation](https://woo.com/document/cart-checkout-blocks-status/#section-6) the notice gets displayed correctly.

Thank you!

**Comment by opr** [comment]
@lovec741 I'm taking a look into this in order to start a discussion about these notices. I am finding your reproduction steps hard to follow as I'm unfamiliar with the webhooks system in WooCommerce.

Could you provide clearer/easier to follow instructions so I can reproduce your issue? Providing some minimal code would be really helpful too

**Comment by opr** [comment]
@lovec741 Hi - just to let you know, I opened a discussion about this issue: https://github.com/woocommerce/woocommerce/discussions/46489 it would be good to hear your thoughts there.

Thanks!
